server/src/main/java/org/elasticsearch/action/bulk/BulkRequestParser.java

```diff
@@ -154,53 +131,228 @@ public void parse(
+        private final Boolean defaultRequireAlias;
+        private final Boolean defaultRequireDataStream;
+        private final Boolean defaultListExecutedPipelines;
+        private final boolean allowExplicitIndex;
+
+        private final XContentType xContentType;
+        private final byte marker;
+        private final BiConsumer<IndexRequest, String> indexRequestConsumer;
+        private final Consumer<UpdateRequest> updateRequestConsumer;
+        private final Consumer<DeleteRequest> deleteRequestConsumer;
+
+        private int incrementalFromOffset = 0;
+        private int line = 0;
+
+        private DocWriteRequest<?> currentRequest = null;
+        private String currentType = null;
+        private String currentPipeline = null;
+        private boolean currentListExecutedPipelines = false;
+
+        private IncrementalParser(
+            @Nullable String defaultIndex,
+            @Nullable String defaultRouting,
+            @Nullable FetchSourceContext defaultFetchSourceContext,
+            @Nullable String defaultPipeline,
+            @Nullable Boolean defaultRequireAlias,
+            @Nullable Boolean defaultRequireDataStream,
+            @Nullable Boolean defaultListExecutedPipelines,
+            boolean allowExplicitIndex,
+            XContentType xContentType,
+            BiConsumer<IndexRequest, String> indexRequestConsumer,
+            Consumer<UpdateRequest> updateRequestConsumer,
+            Consumer<DeleteRequest> deleteRequestConsumer
+        ) {
+            this.defaultIndex = defaultIndex;
+            this.defaultRouting = defaultRouting;
+            this.defaultFetchSourceContext = defaultFetchSourceContext;
+            this.defaultPipeline = defaultPipeline;
+            this.defaultRequireAlias = defaultRequireAlias;
+            this.defaultRequireDataStream = defaultRequireDataStream;
+            this.defaultListExecutedPipelines = defaultListExecutedPipelines;
+            this.allowExplicitIndex = allowExplicitIndex;
+            this.xContentType = xContentType;
+            this.marker = xContentType.xContent().bulkSeparator();
+            this.indexRequestConsumer = indexRequestConsumer;
+            this.updateRequestConsumer = updateRequestConsumer;
+            this.deleteRequestConsumer = deleteRequestConsumer;
+        }
+
+        public int parse(BytesReference data, boolean lastData) throws IOException {
+            int from = 0;
+            int consumed = 0;
+
+            while (true) {
+                int nextMarker = findNextMarker(marker, incrementalFromOffset, data, lastData == false);
+                if (nextMarker == -1) {
+                    incrementalFromOffset = data.length() - consumed;
+                    break;
+                }
+                incrementalFromOffset = nextMarker + 1;
+                line++;
+
+                if (currentRequest == null) {
+                    if (parseActionLine(data, from, nextMarker)) {
```

Comment:
if you fail to parse action line do you discard remaining document line?

Reply:
The parser cannot be reused if it throws an error. I have added a try catch to make this explicit. It also doesn't hold any resources of its own to release.